lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,181 @@
+package graphql.nadel.validation
+
+import graphql.Scalars
+import graphql.nadel.dsl.RemoteArgumentDefinition
+import graphql.nadel.dsl.UnderlyingServiceHydration
+import graphql.nadel.engine.util.isList
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.nadel.engine.util.unwrapOne
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
+        /*
+*  Checks the type of a hydration argument derived from either a $source field output type or $argument field input
+* argument type against the type of an actor field argument to see if they match
+* will also allow for cases of batched hydration eg ID! -> [ID]
+*/
+        fun getHydrationInputErrors(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration
+        ): NadelSchemaValidationError? {
+
+            //need to check null compatibility for inner types (as this is recursive)
+            if (actorFieldArgType.isNonNull && !hydrationSourceFieldType.isNonNull) {
+                // source must be at least as strict as field argument
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                )
+            }
+
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull()
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            // check the actual type
+            if (unwrappedHydrationSourceFieldType is GraphQLScalarType && unwrappedActorFieldArgType is GraphQLScalarType) {
+                return validateScalarArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
+            }
+            else if (unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList) {
+                return validateListArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
+            }
+            else if (unwrappedHydrationSourceFieldType is GraphQLObjectType && unwrappedActorFieldArgType is GraphQLInputObjectType) {
+                return validateInputObjectArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
```

Comment:
```suggestion
                return validateInputObjectArg(unwrappedHydrationSourceFieldType, unwrappedActorFieldArgType, parent, overallField, remoteArg, hydration)
```

You should pass in the smart casted objects instead of passing in the original values then manually unwrapping and casting them.